src/modulus_convert.rs

```diff
@@ -0,0 +1,148 @@
+use crate::error::BoxError;
+use crate::field::Field;
+use crate::helpers::ring::Ring;
+use crate::prss::Participant;
+use crate::replicated_secret_sharing::ReplicatedSecretSharing;
+use crate::securemul::{ProtocolContext, SecureMul};
+use bit_vec::BitVec;
+
+pub enum HelperIdentity {
+    H1,
+    H2,
+    H3,
+}
+
+pub struct RandomShareGenerationHelper<'a> {
+    rng: &'a Participant,
+    identity: HelperIdentity,
+}
+
+impl<'a> RandomShareGenerationHelper<'a> {
+    #[must_use]
+    pub fn new(rng: &'a Participant, identity: HelperIdentity) -> Self {
+        Self { rng, identity }
+    }
+
+    #[must_use]
+    pub fn split_binary<T: Field>(
+        &self,
+        left: bool,
+        right: bool,
+    ) -> (
+        ReplicatedSecretSharing<T>,
+        ReplicatedSecretSharing<T>,
+        ReplicatedSecretSharing<T>,
+    ) {
+        let left = u128::from(left);
+        let right = u128::from(right);
+        match self.identity {
+            HelperIdentity::H1 => (
```

Comment:
I am not sure if it is possible, but it would be easier if each helper does not have an identity but only knows about its left and right peer. In that case it could only generate share for party on the left: always (0, 0), its own share (X_i, 0) and share for helper on the right (0, X_i+1).

Reply:
Unfortunately, it *does* need to know its identity. This particular function here is a perfect example of a time when each of the three helpers needs to do something different, and the optimization which reduces the communication overhead from 6 to 3 is another.

Reply:
This section would be needed for sort also to calculate 1 - x, can we extract this in a function like before?

Reply:
One more example of this would be to get a share of 1 for each helper -- this is needed in sort protocol